### ISSUE TITLE
description of `Data` is not up to date with rest of design

### ISSUE THREAD

**Issue by zygoloid** [carbon-language/carbon-lang#2465]
In [the classes design](https://github.com/carbon-language/carbon-lang/blob/trunk/docs/design/classes.md#nominal-data-classes), we have:

> ### Nominal data classes
>
> We will mark [data classes](https://github.com/carbon-language/carbon-lang/blob/trunk/docs/design/classes.md#data-classes) with an impl as Data {} line.
> [...]
> The word Data here refers to an empty interface in the Carbon prologue.

But, in general, we intend for an empty interface to provide no new operations on a type, and this doesn't seem to fit into that model.

It seems to me that the key operation that we need on a `Data` type is the ability to decompose it into a sequence of fields. Perhaps something like:

```
interface Data {
  let Fields:! Tuple;
  fn Decompose[self: Self]() -> Fields;
}
```

... would support the operations that we intend to permit. (Here, I'm imagining a constraint named `Tuple` that tuple types satisfy.)

Ideally we'd be able to supply a default (template) implementation for `Fields` and `Decompose`, and user-defined types would be able to define these operations in other ways if they have custom storage or representation requirements.

**Comment by geoffromer** [comment]
It seems like a lot of fairly plausible use cases for `Decompose` (e.g. hashing) will be performance-sensitive enough that copying the field values into a tuple will be too costly. As an alternative to consider, `Decompose` could take a pointer to a callback with a generic parameter type, and call it on each field.

**Comment by zygoloid** [comment]
I think calling `Decompose` on each field in turn would make it hard to support some of the desired use cases, such as implementing a short-circuiting comparison. Perhaps `Decompose` could be passed all the fields at once?

**Comment by geoffromer** [comment]
Yeah, passing all the fields at once seems like the best option, assuming we do a good job of making `let` parameter-passing efficient.

Alternatively, we could have the callback return whether it should be called again. That could be more efficient in some corners of the tradeoff space, e.g. where there are a lot of fields but the callback typically only needs the first few. However, my hunch is that those cases will be relatively rare.

Stepping back a minute, though, it seems like we're resorting to inversion of control in order to work around the fact that we expect parameters to be dramatically (even asymptotically) more efficient than return values. It might be worth thinking about whether we can address that performance disparity directly.
